src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -539,60 +539,69 @@ protected override void CreateIndicesForNonTypeMembers(ITypeDefinition typeDef)
 
             int typeRowId = _typeDefs.GetRowId(typeDef);
 
+            // First we find the deleted methods, and add them to our dictionary. This is used later when
+            // processing events, properties, methods, and references.
+            var deletedMethods = _changes.GetDeletedMethods(typeDef);
+            if (deletedMethods.Length > 0)
+            {
+                var deletedTypeMembers = ImmutableDictionary.CreateBuilder<IMethodDefinition, DeletedMethodDefinition>(ReferenceEqualityComparer.Instance);
+                foreach (var methodDef in deletedMethods)
+                {
+                    var oldMethodDef = (IMethodDefinition)methodDef.GetCciAdapter();
+                    deletedTypeMembers.Add(oldMethodDef, new DeletedMethodDefinition(oldMethodDef, typeDef, _typesUsedByDeletedMembers));
+                }
+
+                _deletedTypeMembers.Add(typeDef, deletedTypeMembers.ToImmutableDictionary());
+            }
+
             foreach (var eventDef in typeDef.GetEvents(this.Context))
             {
                 if (!_eventMap.Contains(typeRowId))
                 {
                     _eventMap.Add(typeRowId);
                 }
 
-                this.AddDefIfNecessary(_eventDefs, eventDef);
+                var eventChange = _changes.GetChangeForPossibleReAddedMember(eventDef, DefinitionExistsInAnyPreviousGeneration);
+                this.AddDefIfNecessary(_eventDefs, eventDef, eventChange);
+            }
+
+            var deletedEvents = _changes.GetDeletedEvents(typeDef);
+            foreach (var eventDel in deletedEvents)
+            {
+                var oldEventDef = (IEventDefinition)eventDel.GetCciAdapter();
+
+                // Because deleted event information comes from the associated symbol of the deleted accessors, its safe
+                // to assume that everything will be in the dictionary. We wouldn't be here it if wasn't.
+                var deletedMembers = _deletedTypeMembers[typeDef];
```

Comment:
>var deletedMembers = _deletedTypeMembers[typeDef]; [](http://example.com/codeflow?start=16&length=50)

It looks like we use this multiple times after this point. If the dictionary is not modified below here, consider getting this dictionary value once, outside of this loop, and reusing.